server/tm_cmds.go

```diff
@@ -115,3 +122,130 @@ func VersionCmd() *cobra.Command {
 		},
 	}
 }
+
+// QueryBlocksByEventsCmd returns a command to search through blocks by events.
+func QueryBlocksByEventsCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "blocks",
+		Short: "Query for paginated blocks that match a set of events",
+		Long: `Search for blocks that match the exact given events where results are paginated.
+The events query is directly passed to Tendermint's RPC BlockSearch method and must
+conform to Tendermint's query syntax.
+Please refer to each module's documentation for the full set of events to query
+for. Each module documents its respective events under 'xx_events.md'.
+`,
+		Example: fmt.Sprintf(
+			"$ %s query blocks --query \"message.sender='cosmos1...' AND block.height > 7\" --page 1 --limit 30 --order-by ASC",
+			version.AppName,
+		),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+			query, _ := cmd.Flags().GetString(auth.FlagQuery)
+			page, _ := cmd.Flags().GetInt(flags.FlagPage)
+			limit, _ := cmd.Flags().GetInt(flags.FlagLimit)
+			orderBy, _ := cmd.Flags().GetString(auth.FlagOrderBy)
+
+			blocks, err := rpc.QueryBlocksByEvents(clientCtx, page, limit, query, orderBy)
+			if err != nil {
+				return err
+			}
+
+			return clientCtx.PrintProto(blocks)
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+	cmd.Flags().Int(flags.FlagPage, query.DefaultPage, "Query a specific page of paginated results")
+	cmd.Flags().Int(flags.FlagLimit, query.DefaultLimit, "Query number of transactions results per page returned")
+	cmd.Flags().String(auth.FlagQuery, "", "The blocks events query per Tendermint's query semantics")
+	cmd.Flags().String(auth.FlagOrderBy, "", "The ordering semantics (asc|dsc)")
+	_ = cmd.MarkFlagRequired(auth.FlagQuery)
+
+	return cmd
+}
+
+// QueryBlockCmd implements the default command for a Block query.
+func QueryBlockCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "block --type=[height|hash] [height|hash]",
+		Short: "Query for a committed block by height, hash, or event(s)",
```

Comment:
Maybe we should precise what type of hash is expected (from module consensus or smth).
Otherwise people may try to query app hashes.

Reply:
The only hash that I've found that works is this one for my local chain - does this make sense or would there be others? Also, how would you denote this so that a user would understand what hash is required?```
...
last_block_id:
    hash: 7nI4VVV+uyrFLP2aD7O4PalBLewP8A+I5zJIFLsB8Y0=
...
last_commit:
  block_id:
    hash: 7nI4VVV+uyrFLP2aD7O4PalBLewP8A+I5zJIFLsB8Y0=
...
```